### PR TITLE
Add gcc configuration to jenkins

### PR DESCRIPTION
**Pull request by Rombur**

### CODE REVIEW

File: .jenkins/continuous.groovy [hunk 1]
```diff
@@ -275,6 +275,66 @@ pipeline {
                     }
                 }
 
+                stage('gcc') {
+                    agent {
+                        dockerfile {
+                            filename "Dockerfile"
+                            dir "docker"
+                            additionalBuildArgs '--build-arg BASE=gcc:12.2.0 --build-arg KOKKOS_OPTIONS="-DCMAKE_CXX_EXTENSIONS=OFF -DKokkos_ENABLE_OPENMP=ON -DCMAKE_CXX_COMPILER=g++ -DCMAKE_PREFIX_PATH=/usr/lib/x86_64-linux-gnu"'
```
Comment (masterleinad) [inline .jenkins/continuous.groovy:283]: Why do you need to add `CMAKE_CXX_EXTENSIONS=OFF` and specify `CMAKE_PREFIX_PATH`?

File: .jenkins/continuous.groovy [hunk 2]
```diff
@@ -275,6 +275,66 @@ pipeline {
...
+                        sh 'ccache --zero-stats'
+                        sh 'rm -rf build && mkdir -p build'
+                        dir('build') {
+                            sh '''
+                                cmake \
+                                    -D CMAKE_INSTALL_PREFIX=$ARBORX_DIR \
+                                    -D CMAKE_BUILD_TYPE=Debug \
+                                    -D CMAKE_CXX_COMPILER_LAUNCHER=ccache \
+                                    -D CMAKE_CXX_COMPILER=g++ \
+                                    -D CMAKE_CXX_EXTENSIONS=OFF \
```
Comment (masterleinad) [inline .jenkins/continuous.groovy:298]: Is this necessary?

File: docker/Dockerfile [hunk 3]
```diff
@@ -120,7 +120,7 @@ RUN BOOST_VERSION=1.75.0 && \
 # Install Google Benchmark support library
 ENV BENCHMARK_DIR=/opt/benchmark
 RUN SCRATCH_DIR=/scratch && mkdir -p ${SCRATCH_DIR} && cd ${SCRATCH_DIR} && \
-    git clone https://github.com/google/benchmark.git -b v1.4.1 && \
+    git clone https://github.com/google/benchmark.git -b v1.6.1 && \
```
Comment (masterleinad) [inline docker/Dockerfile:123]: Why did you need to update?

File: examples/molecular_dynamics/example_molecular_dynamics.cpp [hunk 4]
```diff
@@ -148,7 +150,8 @@ int main(int argc, char *argv[])
               KokkosExt::ArithmeticTraits::infinity<decltype(rsq)>::value;
           if (rsq < cutoff_sq_ij)
           {
-            auto const r2inv = static_cast<decltype(rsq)>(1) / rsq;
+            auto const r2inv =
+                static_cast<std::remove_const_t<decltype(rsq)>>(1) / rsq;
```
Comment (masterleinad) [inline examples/molecular_dynamics/example_molecular_dynamics.cpp:154]: Why was casting `1` to `const float` problematic?

File: test/tstDistributedTree.cpp [hunk 5]
```diff
@@ -185,12 +185,11 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(empty_tree, DeviceType, ARBORX_DEVICE_TYPES)
...
   {
-    ARBORX_TEST_QUERY_TREE(
-        ExecutionSpace{}, tree,
-        makeIntersectsSphereQueries<DeviceType>({
-            {{{(double)comm_rank, 0., 0.}}, (double)comm_size},
-        }),
-        make_reference_solution<PairIndexRank>({}, {0, 0}));
+    ARBORX_TEST_QUERY_TREE(ExecutionSpace{}, tree,
+                           makeIntersectsSphereQueries<DeviceType>({
+                               {{{(float)comm_rank, 0., 0.}}, (float)comm_size},
```
Comment (masterleinad) [inline test/tstDistributedTree.cpp:190]: If the types need to match, I would have expected that you also need to use `0.f` instead of `0.` (and I would prefer it for consistency).
```suggestion
                               {{{(float)comm_rank, 0.f, 0.f}}, (float)comm_size},
```

File: .jenkins/continuous.groovy (hunk 1 above)
Comment (Rombur) [reply .jenkins/continuous.groovy:283]: I just copied from the clang stage

File: docker/Dockerfile (hunk 3 above)
Comment (Rombur) [reply docker/Dockerfile:123]: Because version 1.4 does not compile with newer version of gcc

File: examples/molecular_dynamics/example_molecular_dynamics.cpp (hunk 4 above)
Comment (Rombur) [reply examples/molecular_dynamics/example_molecular_dynamics.cpp:154]: I kept getting a warning that the const qualifer was being ignored.

File: docker/Dockerfile (hunk 3 above)
Comment (aprokop) [reply docker/Dockerfile:123]: Not sure I understand. Does it compile with an older version of gcc, but not with a newer?

**Review by aprokop — APPROVED** [review]

**Comment by masterleinad** [comment]
```
file included from /scratch/kokkos/core/src/Kokkos_OpenMP.hpp:222,
                 from /scratch/kokkos/core/src/decl/Kokkos_Declare_OPENMP.hpp:49,
                 from /scratch/kokkos/build/KokkosCore_Config_DeclareBackend.hpp:47,
                 from /scratch/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /scratch/kokkos/core/src/impl/Kokkos_Core.cpp:45:
/scratch/kokkos/core/src/OpenMP/Kokkos_OpenMP_Exec.hpp: In static member function 'static void Kokkos::OpenMP::partition_master(const F&, int, int)':
/scratch/kokkos/core/src/OpenMP/Kokkos_OpenMP_Exec.hpp:182:29: warning: 'static void Kokkos::Impl::OpenMPExec::validate_partition(int, int&, int&)' is deprecated [-Wdeprecated-declarations]
  182 |     Exec::validate_partition(prev_instance->m_pool_size, num_partitions,
      |     ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  183 |                              partition_size);
      |                              ~~~~~~~~~~~~~~~
/scratch/kokkos/core/src/OpenMP/Kokkos_OpenMP_Exec.hpp:94:33: note: declared here
   94 |   KOKKOS_DEPRECATED static void validate_partition(const int nthreads,
      |                                 ^~~~~~~~~~~~~~~~~~
[0m[91mIn file included from /scratch/kokkos/core/src/Kokkos_OpenMP.hpp:222,
                 from /scratch/kokkos/core/src/decl/Kokkos_Declare_OPENMP.hpp:49,
                 from /scratch/kokkos/build/KokkosCore_Config_DeclareBackend.hpp:47,
                 from /scratch/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /scratch/kokkos/core/src/impl/Kokkos_ExecPolicy.cpp:45:
/scratch/kokkos/core/src/OpenMP/Kokkos_OpenMP_Exec.hpp: In static member function 'static void Kokkos::OpenMP::partition_master(const F&, int, int)':
/scratch/kokkos/core/src/OpenMP/Kokkos_OpenMP_Exec.hpp:182:29: warning: 'static void Kokkos::Impl::OpenMPExec::validate_partition(int, int&, int&)' is deprecated [-Wdeprecated-declarations]
  182 |     Exec::validate_partition(prev_instance->m_pool_size, num_partitions,
      |     ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  183 |                              partition_size);
      |                              ~~~~~~~~~~~~~~~
/scratch/kokkos/core/src/OpenMP/Kokkos_OpenMP_Exec.hpp:94:33: note: declared here
   94 |   KOKKOS_DEPRECATED static void validate_partition(const int nthreads,
      |                                 ^~~~~~~~~~~~~~~~~~
```

File: docker/Dockerfile (hunk 3 above)
Comment (Rombur) [reply docker/Dockerfile:123]: Yes, because they forgot some includes of the standard library. It happens to work in older version of gcc but in the newer ones, you get an error about missing include

**Review by masterleinad — APPROVED** [review]
Fine with me.

**Comment by aprokop** [comment]
@Rombur please update the wiki's testing page.
